### PR TITLE
Faster `jnp.trapezoid` when `dx` is a scalar

### PR DESCRIPTION
**Pull request by jurasic-pf**
Addressing https://github.com/jax-ml/jax/issues/34915 

**Microbenchmark before**:
`jnp.trapezoid` have the same performance when using `x` or `dx`. For comparison `jnp.sum * dx` is significantly faster.
Image

**Microbenchmark after**:
`jnp.trapezoid` is faster when using `dx` and in line with the performance of `jnp.sum * dx`.
Image

`dx_reduced * (y_arr.sum(-1) - 0.5 * (y_arr[..., 0] + y_arr[..., -1]))`
also seems to be faster than
`dx_reduced * 0.5 * (y_arr[..., 1:] + y_arr[..., :-1]).sum(-1)`

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -6536,6 +6536,10 @@ def trapezoid(y: ArrayLike, x: ArrayLike | None = None, dx: ArrayLike = 1.0,
     else:
       dx_array = moveaxis(diff(x_arr, axis=axis), axis, -1)
   y_arr = moveaxis(y_arr, axis, -1)
+  # Fast path: dx is constant along the integration axis.
+  if dx_array.ndim == 0 or dx_array.shape[-1] == 1:
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:6543]: Some of the existing logic here seems incorrect: if `dx` is specified and is not a scalar, and if `axis != -1`, then `dx` is never transposed to match the reduction dimension of `y_arr`. Would you be willing to fix that while we're here?

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -5988,6 +5979,28 @@ def test_trapezoid(self, yshape, xshape, dtype, dx, axis):
...
                           check_dtypes=False)
 
+  @jtu.sample_product(
+    [dict(yshape=yshape, dx=dx, axis=axis)
+      for yshape, dx, axis in [
+        ((3, 10), np.ones((3, 1), dtype=np.float32), -1),
+        ((10,), np.ones((1,), dtype=np.float32), -1),
+        ((2, 3, 4, 10), np.ones((2, 3, 4, 1), dtype=np.float32), -1),
+        ((2, 3, 4, 10), np.ones((1, 1, 4, 1), dtype=np.float32), -1),
+        ((2, 3, 4, 10), np.ones((2, 1, 1, 1), dtype=np.float32), 1),
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: This test case would have caught the bug if you didn't have so many `1` axes.

File: tests/lax_numpy_test.py
Comment (jurasic-pf) [inline tests/lax_numpy_test.py]: With the old tolerances, this test would fail against numpy, due to different floating point truncation at low precision. In fact the new method is **more accurate** when comparing to the result at e.g. `np.float64` precision:


```python
>>> np.trapezoid(a.astype(np.float16), dx=2.0)
np.float16(4.246)
>>> jnp.trapezoid(a.astype(np.float16), dx=2.0)
Array(4.25, dtype=float16)  # Values are substantially different

>>> np.trapezoid(a.astype(np.float32), dx=2.0)
np.float32(4.24967)  # Because they experience less numerical truncation, and are more accurate.
>>> jnp.trapezoid(a.astype(np.float32), dx=2.0)
Array(4.249671, dtype=float32)
```

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (jurasic-pf) [reply tests/lax_numpy_test.py]: Thanks for pointing that out. I wrote it to test regressions against the old implementation and numpy before reading your comment that `dx` handling was wrong. Should be covered now.

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -6528,6 +6528,11 @@ def trapezoid(y: ArrayLike, x: ArrayLike | None = None, dx: ArrayLike = 1.0,
     y = util.ensure_arraylike('trapezoid', y)
     y_arr, = util.promote_dtypes_inexact(y)
     dx_array = asarray(dx)
+    if dx_array.ndim > 0:
+      # Match y's ndim then moveaxis to match y's rearrangement.
+      dx_array = dx_array.reshape(
+          (1,) * (y_arr.ndim - dx_array.ndim) + dx_array.shape)
+      dx_array = moveaxis(dx_array, axis, -1)
```
Comment (jurasic-pf) [inline jax/_src/numpy/lax_numpy.py:6533]: To address the pre-existing bug with dx handling when `axis != -1`

The broadcasting logic is a bit subtle but matches all the cases I could think of against `numpy`.
Numpy does this handling a lot more verbosely, with `slice` objects, which we can also do if you think this isn't readable enough.

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -5981,13 +5972,43 @@ def test_trapezoid(self, yshape, xshape, dtype, dx, axis):
     args_maker = lambda: [rng(yshape, dtype), rng(xshape, dtype) if xshape is not None else None]
     np_fun = partial(np.trapezoid, dx=dx, axis=axis)
     jnp_fun = partial(jnp.trapezoid, dx=dx, axis=axis)
-    tol = jtu.tolerance(dtype, {np.float16: 2e-3, np.float64: 1e-12,
+    tol = jtu.tolerance(dtype, {np.float16: 4e-3, np.float64: 1e-12,
```
Comment (jurasic-pf) [inline tests/lax_numpy_test.py:5975]: With the old tolerances, this test would fail against numpy, due to different floating point truncation at low precision. In fact the new method is **more accurate** when comparing to the result at e.g. `np.float64` precision:


```python
>>> np.trapezoid(a.astype(np.float16), dx=2.0)
np.float16(4.246)
>>> jnp.trapezoid(a.astype(np.float16), dx=2.0)
Array(4.25, dtype=float16)  # Values are substantially different

>>> np.trapezoid(a.astype(np.float32), dx=2.0)
np.float32(4.24967)  # Because they experience less numerical truncation, and are more accurate.
>>> jnp.trapezoid(a.astype(np.float32), dx=2.0)
Array(4.249671, dtype=float32)
```

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -6528,6 +6528,11 @@ def trapezoid(y: ArrayLike, x: ArrayLike | None = None, dx: ArrayLike = 1.0,
     y = util.ensure_arraylike('trapezoid', y)
     y_arr, = util.promote_dtypes_inexact(y)
     dx_array = asarray(dx)
+    if dx_array.ndim > 0:
+      # Match y's ndim then moveaxis to match y's rearrangement.
+      dx_array = dx_array.reshape(
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py]: Instead of `reshape` here, it would be better to use `expand_dims` (it's a bit more convenient, and also maintains shardings when they're present).

File: jax/_src/numpy/lax_numpy.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/numpy/lax_numpy.py:6533]: Looks good!

**Review by jakevdp — COMMENTED** [review]
Looks great! Just one requested change below.

File: jax/_src/numpy/lax_numpy.py (hunk 5 above)
Comment (jurasic-pf) [reply jax/_src/numpy/lax_numpy.py]: Didn't know about the different sharding behavior - useful!

**Comment by jurasic-pf** [comment]
> Looks great! Just one requested change below.

All changes have been addressed

File: jax/_src/numpy/lax_numpy.py [hunk 6]
```diff
@@ -6528,6 +6528,9 @@ def trapezoid(y: ArrayLike, x: ArrayLike | None = None, dx: ArrayLike = 1.0,
     y = util.ensure_arraylike('trapezoid', y)
     y_arr, = util.promote_dtypes_inexact(y)
     dx_array = asarray(dx)
+    if dx_array.ndim > 0:
+      dx_array = broadcast_to(dx_array, (1,) * (y_arr.ndim - dx_array.ndim) + dx_array.shape)
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py]: Simpler:
```python
dx_array = lax.expand_dims(dx_array, range(y_arr.ndim - dx_arr.ndim))
```

File: tests/lax_numpy_test.py [hunk 7]
```diff
@@ -5981,13 +5972,43 @@ def test_trapezoid(self, yshape, xshape, dtype, dx, axis):
...
+        # slow path: varies along integration axis
+        ((3, 4, 5), (2, 4, 5), 0),
+        ((3, 4, 5), (3, 3, 5), 1),
+      ]
+    ],
+  )
+  @jtu.skip_on_devices("tpu")
+  def test_trapezoid_array_dx(self, yshape, dxshape, axis):
+    rng = jtu.rand_default(self.rng())
+    args_maker = lambda: [rng(yshape, np.float32)]
```
Comment (jakevdp) [inline tests/lax_numpy_test.py]: We should include `dx` in the `args_maker`. The reason for this is that it will confirm that the function works as expected when `dx` is a traced value.

It means that `np_fun` and `jnp_fun` can no longer be defined via `partial`, but you can use `lambda` definitions instead.

Comment (jurasic-pf) [reply tests/lax_numpy_test.py]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 6 above)
Comment (jurasic-pf) [reply jax/_src/numpy/lax_numpy.py]: Done

**Review by jakevdp — COMMENTED** [review]
Looks great! Last request: can you please squash all changes into a single commit? See https://docs.jax.dev/en/latest/contributing.html#single-change-commits-and-pull-requests. Thanks!

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by jurasic-pf** [comment]
> Looks great, thanks!

Thanks for the guidance! Happy to contribute :)
